[TS-886] Undefined fields check option

Toru8P opened the pull request.

Commit:
first implementation

Commit:
readme and version update

Commit:
switched to error

Commit:
common update to 3.32.0
codec update to 4.3.0

Commit:
null check for schema type

Commit:
null check fix for object type inside of array

Commit:
session alias and protocol loss fix

Commit:
session alias of parsed encoded header loss fix

Commit:
null default values removed from result message

Commit:
Readme for release notes

Commit:
Refactoring

Commit:
Refactoring, fixed case method sensitive search

Commit:
readme update

Commit:
Charset support inside of type, multiprotocol update (json, binary)

Commit:
Fixed protocol check, package refactoring, tests fixed for protocol

Commit:
message type for decoded messages

Commit:
Any number support

Commit:
Test updated for BigDecimal types, http as supported protocol fixed

Commit:
Request and Response extracted from encode pipeline

Commit:
Protocol http added to created request and response header messages

cordwelt commented on README.md:
```diff
@@ -140,7 +140,7 @@ Result of decode:
 ### Codec configs:
 
-Config contain two main types of options.
+* warnUndefinedFields - Enable or Disable warnings for all undefined fields inside object structures, true by default.
```
```suggestion
* checkUndefinedFields - Enable or Disable warnings for all undefined fields inside object structures, true by default.
```

Commit:
Method error fixed

Commit:
trace for encoded

Commit:
More log messages

Commit:
header message missed fix

Commit:
additional log messages

Commit:
additional log messages

Commit:
additional log messages

cordwelt commented on README.md:
```diff
@@ -47,7 +47,7 @@ Path that contains parameters inside will be converted as:
 All messages processing by encode must contain those types.
 
-Decode accepts only raw messages, all parsed will be passed.
+Encode accepts only parsed messages, all raw will be passed.
```
```suggestion
Encode accepts only parsed messages, raw ones will be skipped.
```

cordwelt commented on README.md:
```diff
@@ -163,10 +163,21 @@ Config contain two main types of options.
+
+### 0.2.0
+
++ Feature: check undefined fields and throws errors if they are found
```
```suggestion
+ Feature: check for undefined fields and throw errors if they are found
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -103,12 +108,7 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
-            }
-
-            if (message.message.metadata.run { protocol.isNotEmpty() && protocol != PROTOCOL }) {
+            if (message.kindCase != MESSAGE || message.message.metadata.run { protocol.isNotEmpty() && !PROTOCOLS.contains(protocol) } || message.message.messageType == REQUEST_MESSAGE || message.message.messageType == RESPONSE_MESSAGE) {
```
I'd suggest to rollback to multiple `if`'s instead and reuse existing `parsedMessage`, `metadata` and `messageType` variables like this:

```
if (!message.hasMessage()) {
    builder.addMessages(message)
    continue
}

val parsedMessage = message.message
val metadata = parsedMessage.metadata
val messageType = metadata.messageType

if(messageType == REQUEST_MESSAGE || messageType == RESPONSE_MESSAGE || metadata.protocol.run { isNotEmpty() && this !in PROTOCOLS }) {
    builder.addMessages(message)
    continue
}
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -117,38 +117,54 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
-            runCatching {
+            builder += header
+
+            LOGGER.trace { "Created header message for ${parsedMessage.messageType}: ${header.messageType}" }
```
```
builder += createHeaderMessage(container, parsedMessage).apply {
    if (parsedMessage.hasParentEventId()) parentEventId = parsedMessage.parentEventId
    sessionAlias = parsedMessage.sessionAlias
    metadataBuilder.putAllProperties(parsedMessage.metadata.propertiesMap)
    LOGGER.trace { "Created header message for ${parsedMessage.messageType}: ${this.messageType}" }
}
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -117,38 +117,54 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
-    private fun encodeBody(container: HttpRouteContainer, message: Message): RawMessage? = container.body?.let { messageSchema ->
+    private fun encodeBody(container: HttpRouteContainer, message: Message): RawMessage? {
+        val messageSchema = container.body ?: run {
+            LOGGER.trace { "Body wasn't found for message: ${message.messageType}" }
```
Why trace? Is it expected to happen a lot?

Toru8P replied:
True, debug is reasonable

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -180,47 +196,64 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
         val body = rawMessage.body
 
-        val bodyFormat = header.getList(HEADERS_FIELD)?.first { it.messageValue.getString("name") == "Content-Type" }?.messageValue?.getString("value") ?: "null"
+        val bodyFormat = header.getList(HEADERS_FIELD)?.first { it.messageValue.getString("name") == "Content-Type" }?.messageValue?.getString("value")?.extractType() ?: "null"
```
```suggestion
        val bodyFormat = header.getList(HEADERS_FIELD)
            ?.first { it.messageValue.getString("name") == "Content-Type" }
            ?.messageValue
            ?.getString("value")
            ?.extractType() ?: "null"
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -180,47 +196,64 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
+        val method: String
+        var code = ""
+
+        val pairFound: Pair<UriPattern, PathItem>
```
Maybe use data class instead?

Toru8P replied:
Right now i cant see scenario when we need more than two values for this case

cordwelt replied:
The point is not about the amount values, but about readability because `pairFound.first` / `pairFound.second` doesn't give you any info about what's inside

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -180,47 +196,64 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
-        SchemaWriter.instance.traverse(visitor, messageSchema)
+        schemaWriter.traverse(visitor, messageSchema)
         return visitor.getResult().apply {
             parentEventId = rawMessage.parentEventId
```
`if(rawMessage.hasParentEventId()) parentEventId = rawMessage.parentEventId`

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -180,47 +196,64 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
         when (this) {
             is ResponseContainer -> metadata.apply {
-                putProperties(METHOD_PROPERTY, method)
+                method?.let { putProperties(METHOD_PROPERTY, it.uppercase(Locale.getDefault())) }
```
```suggestion
                method?.let { putProperties(METHOD_PROPERTY, it.uppercase()) }
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodec.kt:
```diff
@@ -180,47 +196,64 @@ class OpenApiCodec(private val dictionary: OpenAPI) : IPipelineCodec {
             }
             is RequestContainer -> metadata.apply {
-                putProperties(METHOD_PROPERTY, method)
+                method?.let { putProperties(METHOD_PROPERTY, it.uppercase(Locale.getDefault())) }
```
```suggestion
                method?.let { putProperties(METHOD_PROPERTY, it.uppercase()) }
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/OpenApiCodecFactory.kt:
```diff
@@ -26,13 +26,14 @@ import io.swagger.parser.OpenAPIParser
     override val settingsClass: Class<out IPipelineCodecSettings> = OpenApiCodecSettings::class.java
-    override val protocol: String = PROTOCOL
+    override val protocols: Set<String>
+        get() = PROTOCOLS
```
```suggestion
    override val protocols: Set<String> = PROTOCOLS
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/utils/JsonUtils.kt:
```diff
@@ -25,19 +25,26 @@ import com.fasterxml.jackson.databind.JsonNode
+    } else {
+        node
     }
 }
```
```suggestion
fun JsonNode.getRequiredField(fieldName: String, required: Boolean): JsonNode? = get(fieldName).also { node ->
    check(!required || node != null && node !is NullNode) { "Field [$fieldName] is required for json [${asText()}]" }
    if(node is NullNode) return null
```

Toru8P replied:
But node can be null as well as NullNode, does we really need to check twice Null and NullNode status?

cordwelt replied:
We only check twice for `NullNode` in some cases. I'm not against original implementation, just wanted to make it a bit shorter

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/utils/OpenApiUtils.kt:
```diff
@@ -59,4 +59,20 @@ fun <T> Schema<*>.checkEnum(value: T?, name: String) {
     }
+}
+
+fun String.extractType() = this.split(";")[0].trim()
```
```suggestion
fun String.extractType() = substringBefore(';').trim()
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/utils/OpenApiUtils.kt:
```diff
@@ -59,4 +59,20 @@ fun <T> Schema<*>.checkEnum(value: T?, name: String) {
+    val schema = if (code!=null) {
+        methodRootSchema?.responses?.get(code)?.content?.get(bodyFormat)?.schema
+    } else {
+        methodRootSchema?.requestBody?.content?.get(bodyFormat)?.schema
```
```suggestion
    val schema = when (code) {
        null -> methodRootSchema?.requestBody?.content?.get(bodyFormat)?.schema
        else -> methodRootSchema?.responses?.get(code)?.content?.get(bodyFormat)?.schema
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/SchemaWriter.kt:
```diff
@@ -27,9 +27,10 @@ import io.swagger.v3.parser.util.SchemaTypeUtil.INTEGER_TYPE
 
 
-class SchemaWriter private constructor(private val openApi: OpenAPI) {
+class SchemaWriter constructor(private val openApi: OpenAPI, private val exceptionUndefined: Boolean = true) {
```
```suggestion
class SchemaWriter constructor(private val openApi: OpenAPI, private val failOnUndefined: Boolean = true) {
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/SchemaWriter.kt:
```diff
@@ -43,6 +44,12 @@ class SchemaWriter private constructor(private val openApi: OpenAPI) {
                 requireNotNull(schema.properties) {"Properties in object are required: $schema"}
+                if (exceptionUndefined) {
+                    schemaVisitor.visitUndefinedFields(schema.properties.keys)?.let {
+                        if (it.isNotEmpty()) error("Undefined fields was found inside of ${schema.name}: ${it.joinToString(", ")}")
```
```suggestion
                        check(it.isEmpty()) { "Undefined fields were found inside of ${schema.name}: ${it.joinToString()}" }
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/SchemaWriter.kt:
```diff
@@ -80,6 +87,9 @@ class SchemaWriter private constructor(private val openApi: OpenAPI) {
                         }
+                        "-" -> {
+                            visitor.visit(name, property.default as? BigDecimal, property, required)
+                        }
```
I'd suggest removing braces from all `when` entries

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/SchemaVisitor.kt:
```diff
@@ -16,27 +16,32 @@
     abstract fun visitLongCollection(fieldName: String, defaultValue: List<Long>?, fldStruct: ArraySchema, required: Boolean = false)
-    abstract fun visitObjectCollection(fieldName: String, defaultValue: List<Any>?, fldStruct: ArraySchema, required: Boolean = false)
+    abstract fun visitObjectCollection(fieldName: String, defaultValue: List<Any>?, fldStruct: ArraySchema, required: Boolean = false, schemaWriter: SchemaWriter)
+    abstract fun visitUndefinedFields(fields: MutableSet<String>): Set<String>?
```
Maybe rename to `getUndefinedFields`?

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/SchemaVisitor.kt:
```diff
@@ -16,27 +16,32 @@
+    abstract fun visitObjectCollection(fieldName: String, defaultValue: List<Any>?, fldStruct: ArraySchema, required: Boolean = false, schemaWriter: SchemaWriter)
+    abstract fun visitUndefinedFields(fields: MutableSet<String>): Set<String>?
 
     abstract fun getResult(): ToType
```
Maybe convert to property?

Toru8P replied:
Isn't convertation process is too heavy to be used in property?

cordwelt replied:
By conversion I meant to replace:

> fun getResult(): ToType

with:

> val result: ToType

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/DecodeJsonArrayVisitor.kt:
```diff
@@ -90,17 +96,24 @@ class DecodeJsonArrayVisitor(override val from: ArrayNode) : DecodeVisitor<Array
 
-    override fun visitObjectCollection(fieldName: String, defaultValue: List<Any>?, fldStruct: ArraySchema, required: Boolean) {
+    override fun visitBigDecimalCollection(fieldName: String, defaultValue: List<BigDecimal>?, fldStruct: ArraySchema, required: Boolean) {
+        rootMessage.addField(fieldName, from.map { it.validateAsBigDecimal() })
```
> { it.validateAsBigDecimal() }

Maybe convert to reference?

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/DecodeJsonObjectVisitor.kt:
```diff
@@ -121,19 +130,29 @@ class DecodeJsonObjectVisitor(override val from: ObjectNode) : DecodeVisitor<Obj
         }
     }
 
+    override fun visitUndefinedFields(fields: MutableSet<String>): Set<String> = this.from.fieldNames().asSequence().toMutableSet().apply { removeAll(fields) }
```
```suggestion
    override fun visitUndefinedFields(fields: MutableSet<String>): Set<String> = from.fieldNames().asSequence().filterNot { it in fields }.toSet()
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/EncodeJsonObjectVisitor.kt:
```diff
@@ -23,64 +23,94 @@ import com.exactpro.th2.codec.openapi.writer.SchemaWriter
-        val value = from.getRequiredField(fieldName, required)?.getString()?.toBoolean()
-        fldStruct.checkEnum(value, fieldName)
-        rootNode.put(fieldName, value ?: defaultValue)
+        from.getRequiredField(fieldName, required)?.getString()?.toBoolean()?.let { value ->
+            fldStruct.checkEnum(value, fieldName)
+            rootNode.put(fieldName, value)
+        } ?: defaultValue?.let {
+            rootNode.put(fieldName, defaultValue)
+        }
```
Can be extracted into a separate method. Also I'd suggest renaming `getRequiredField` method to `getField` since its current name implies that it always checks field presence

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/EncodeJsonObjectVisitor.kt:
```diff
@@ -119,20 +149,30 @@ class EncodeJsonObjectVisitor(override val from: Message) : EncodeVisitor<Messag
 
-    override fun visitObjectCollection(fieldName: String, defaultValue: List<Any>?, fldStruct: ArraySchema, required: Boolean) {
+    override fun visitBigDecimalCollection(fieldName: String, defaultValue: List<BigDecimal>?, fldStruct: ArraySchema, required: Boolean) {
+        from.getRequiredField(fieldName, required)?.getList()?.let { values ->
+            rootNode.putArray(fieldName).putAll<BigDecimal>(values)
+        }
```
Why `defaultValue` is ignored?

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/EncodeJsonObjectVisitor.kt:
```diff
@@ -119,20 +149,30 @@ class EncodeJsonObjectVisitor(override val from: Message) : EncodeVisitor<Messag
         }?.run(rootNode.putArray(fieldName)::addAll)
     }
 
+    override fun visitUndefinedFields(fields: MutableSet<String>): Set<String> = this.from.fieldsMap.keys.minus(fields)
```
```suggestion
    override fun visitUndefinedFields(fields: MutableSet<String>): Set<String> = from.fieldsMap.keys - fields
```

Commit:
big decimal without scientific format on decode

Commit:
Review refactoring
Message type for empty (body) http packages will be generated only if body not required

Commit:
Readme updated

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/utils/CommonUtils.kt:
```diff
@@ -20,7 +20,15 @@ import com.exactpro.th2.common.grpc.Message
+ * @param fieldName name of extracted field
+ * @param required check if value is required, used only if extracted value was null
+ */
+fun Message.getField(fieldName: String, required: Boolean): Value? = this@getField.getField(fieldName).apply {
```
```suggestion
fun Message.getField(fieldName: String, required: Boolean): Value? = this[fieldName].apply {
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/utils/JsonUtils.kt:
```diff
@@ -30,14 +30,25 @@ import java.math.BigDecimal
     String::class -> values.forEach { add(it.getString()) }
-    Float::class, Double::class -> values.forEach { add(it.getBigDecimal()) }
-    Boolean::class -> values.forEach { add(it.getString()!!.toBoolean()) }
+    Float::class, Double::class -> values.forEach { add(it.getFloat()) }
```
I highly advise against using `float` type. Also what's wrong with converting `float` and `double` to `BigDecimal`?

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/utils/JsonUtils.kt:
```diff
@@ -30,14 +30,25 @@ import java.math.BigDecimal
+inline fun <reified T> ArrayNode.putAll(values: List<T>) = when (T::class) {
+    Int::class -> values.forEach { add(it as Int) }
+    String::class -> values.forEach { add(it as String) }
+    Float::class, Double::class -> values.forEach { add(it as Float) }
```
Same here

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/DecodeJsonArrayVisitor.kt:
```diff
@@ -108,10 +95,21 @@ class DecodeJsonArrayVisitor(override val from: ArrayNode) : DecodeVisitor<Array
+                !defaultValue.isNullOrEmpty() -> this.addField(name, defaultValue)
+            }
+        } else {
+            this.addField(name, node.map { it.extract() })
```
```suggestion
            this.addField(name, node.map(extract))
```

cordwelt commented on src/main/kotlin/com/exactpro/th2/codec/openapi/writer/visitors/json/EncodeJsonArrayVisitor.kt:
```diff
@@ -129,4 +86,12 @@ class EncodeJsonArrayVisitor(override val from: Message) : EncodeVisitor<Message
     }
+
+    private inline fun <reified T> ArrayNode.putListFrom(message: Message, name: String, defaultValue: List<T>?, required: Boolean) {
+        message.getField(name, required)?.getList()?.let { values ->
+            this.putAll<T>(values)
+        } ?: defaultValue?.let { list ->
+            this.putAll(list)
+        }
```
```suggestion
        message.getField(name, required)?.getList()?.let { putAll<T>(it) } ?: defaultValue?.let(::putAll)
```

Commit:
codec version up to 4.6.1

Commit:
Review refactoring
